src/assign.rs

```diff
@@ -313,26 +313,160 @@ mod json {
+                },
+                Test {
+                    data: json!({"foo": "bar"}),
+                    ptr: "/foo/-",
+                    value: json!("baz"),
+                    expected_data: json!({"foo": ["baz"]}),
+                    expected_result: Ok(Some(json!("bar"))),
+                },
+                Test {
+                    data: json!({}),
+                    ptr: "/0",
+                    value: json!("foo"),
+                    expected_data: json!({"0": "foo"}),
+                    expected_result: Ok(None),
+                },
+                Test {
+                    data: json!(null),
+                    ptr: "/1",
+                    value: json!("foo"),
+                    expected_data: json!({"1": "foo"}),
+                    expected_result: Ok(Some(json!(null))),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/0",
+                    expected_data: json!(["foo"]),
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/0",
+                    expected_data: json!(["foo"]),
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/-",
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                    expected_data: json!(["foo"]),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/1",
+                    value: json!("foo"),
+                    expected_result: Err(AssignError::OutOfBounds {
+                        offset: 0,
+                        source: OutOfBoundsError{
+                            index: 1,
+                            length: 0,
+                        },
+                    }),
+                    expected_data: json!([]),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/a",
+                    value: json!("foo"),
+                    expected_result: Err(AssignError::FailedToParseIndex {
+                        offset: 0,
+                        source: json::ParseIndexError { source: usize::from_str("foo").unwrap_err() },
+                    }),
+                    expected_data: json!([]),
+                },
+            ];
 
-            let tests = [(json!({}), "/foo", json!("bar"))];
-            for (mut data, ptr, value) in tests {
+            for (i,Test {
+                mut data,
+                ptr,
+                value,
+                expected_data,
+                expected_result: expected_replaced,
+            }) in tests.into_iter().enumerate()
+            {
                 let ptr = PointerBuf::from_str(ptr).unwrap();
```

Comment:
Missed this on the earlier version, but why `PointerBuf`? It doesn't look like it requires mutation.

Reply:
It doesn't. This was another one of those that I think was either a `Pointer` before or I just spaced out and wrote `PointerBuf`.

Thanks for catching it. I got this and the rest.